packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -300,6 +300,60 @@ function warnIfInvokedDuringConstruction(vm: VM, methodOrPropName: string) {
     }
 }
 
+// List of properties on ElementInternals that are formAssociated can be found in the spec:
+// https://html.spec.whatwg.org/multipage/custom-elements.html#form-associated-custom-elements
+const formAssociatedProps = new Set([
+    'setFormValue',
+    'form',
+    'setValidity',
+    'willValidate',
+    'validity',
+    'validationMessage',
+    'checkValidity',
+    'reportValidity',
+    'labels',
+]);
+
+// Wrap all ElementInternal objects in a proxy to prevent form association when `formAssociated` is not set on an LWC.
+// This is needed because the 1UpgradeableConstructor1 always sets `formAssociated=true`, which means all
+// ElementInternal objects will have form-associated properties set when an LWC is placed in a form.
+// We are doing this to guard against customers taking a dependency on form elements being associated to ElementInternals
+// when 'formAssociated' has not been set on the LWC.
+function createElementInternalsProxy(
+    elementInternals: ElementInternals,
+    isFormAssociated: boolean
+) {
+    const elementInternalsProxy = new Proxy(elementInternals, {
+        set(target, propertyKey, newValue) {
+            const key = typeof propertyKey === 'symbol' ? propertyKey.description : propertyKey;
+            if (key && formAssociatedProps.has(key) && !isFormAssociated) {
+                //Note this error message mirrors Chrome and Firefox error messages, in Safari the error is slightly different.
+                throw new DOMException(
+                    `Failed to execute '${key}' on 'ElementInternals': The target element is not a form-associated custom element.`
+                );
+            }
+            return Reflect.set(target, propertyKey, newValue);
```

Comment:
I'm not too familiar with the `Reflect` API; this may warrant a comment.

Reply:
I really used this for convenience, the [`Reflect.get`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Reflect/get) does the same things as `target[propertyKey]`. 

The convenience here is that if we try to do `target[propertyKey]` typescript throws an error saying "No index signature with a parameter of type 'string' was found on type 'ElementInternals'".

I think I'd need to define a key type to be able to index into the `ElementInternals` object, whereas with `Reflect` I can call the API to get the value back.

Let me know if I should add some comments around it though!

Reply:
That's fine, makes sense to me. TIL!